A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,225 @@
+[C++](https://github.com/grpc/grpc/blob/4567af504ed53cc8398e53bf1efd23f753d43bb8/src/core/ext/filters/client_channel/lb_policy.h#L178-L201),
+[Go](https://github.com/grpc/grpc-go/blob/01ed64857e3146000ec99cdea4f2932204f17cdd/balancer/balancer.go#L256-L269)
+), the picker should add a
+callback for atomically decrementing the subchannel counter once the RPC finishes (regardless of `Status` code).
+This approach closely resembles the implementation for outstanding request counters already present in
+the `cluster_impl_experimental` policy (
+[Java](https://github.com/grpc/grpc-java/blob/0000cba665c69958355b639474c7387d98afcc79/xds/src/main/java/io/grpc/xds/ClusterImplLoadBalancer.java#L282-L379),
+[C++](https://github.com/grpc/grpc/blob/79d684529d9db60aa2c5e82deb43e297a6818cfb/src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_impl.cc#L285-L354),
+[Go](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/xds/internal/balancer/clusterimpl/picker.go#L102-L191)
+).
+This approach entails some degree of raciness which will be discussed later
+(See [Outstanding request counter raciness](#outstanding-request-counter-raciness)).
+
+### Temporary environment variable protection
+
+During initial development, this feature will be enabled via the
+`GRPC_XDS_EXPERIMENTAL_ENABLE_LEAST_REQUEST` environment variable.  This
+environment variable protection will be removed once the feature has
+proven stable.
+
+## Rationale
+
+### Support for Weighted Least Request
+
+While also supporting the "*all weights not equal*" (Weighted Least Request) case
+[from Envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#weighted-least-request)
+would enable more advanced use cases, there are some missing pieces in gRPC that
+are required for that to be technically feasible.
+For example, gRPC ignores the per-endpoint `load_balancing_weight` field as a result
+of the implementation described by
+[gRFC A27: xDS-based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md).
+Later in the future when [ORCA](https://docs.google.com/document/d/1NSnK3346BkBo1JUU3I9I5NYYnaJZQPt8_Z_XCBCI3uA/edit)
+is supported, gRPC may be able to locally provide per-endpoint weights that could be
+used in a weighted least request policy implementation.
+
+### Allow picking same candidate multiple times
+
+One possible downside of the Envoy implementation, is that the random selection will allow a subchannel to
+be sampled multiple times and thereby be unavoidably picked with some probability.
+This means that even a frozen endpoint would never fully be excluded by the picker.
+There exists a
+[closed PR to Envoy](https://github.com/envoyproxy/envoy/pull/11006)
+that attempted to solve this.
+In that PR, there were some questions around whether there would be any significant improvement for >5 endpoints.
+
+### Outstanding request counter raciness
```

Comment:
I think that the new API in C-core that I mentioned above would actually allow you to eliminate most of this raciness -- or at least to err on the side of considering an address to have every request you assign to it, even if some of those picks get abandoned.

Reply:
Even though raciness may be eliminated in that particular case I still think it's worth mentioning that some small amount of raciness is theoretically possible.
Since the picker is fully concurrent, any large burst of picks could potentially also cause unfair bursts to individual upstream endpoint(s) which may overload them.
The race window here should be very minor however, this probably isn't a problem in practice.